src/couch_views/src/couch_views_reduce_fdb.erl

```diff
@@ -0,0 +1,238 @@
+        reduce_funs = ViewReduceFuns,
+        id_num = ViewId
+    } = View,
+
+    lists:foldl(fun(ReduceFun, Acc) ->
+        Tree = open_tree(TxDb, Sig, ViewId, ReduceFun),
+        {_, KVSize} = ebtree:full_reduce(Tx, Tree),
+        Acc + KVSize
+    end, 0, ViewReduceFuns).
+
+
+fold(Db, Sig, ViewId, Reducer, GroupLevel, Opts, UserCallback,
+    UserAcc0) ->
+
+    Acc0 = #{
+        user_callback => UserCallback,
+        user_acc => UserAcc0,
+        reducer => Reducer
+    },
+
+    fabric2_fdb:transactional(Db, fun(TxDb) ->
+        #{
+            tx := Tx
+        } = TxDb,
+
+        StartKey = fabric2_util:get_value(start_key, Opts),
+        EndKey = fabric2_util:get_value(end_key, Opts),
+
+        Tree = open_tree(TxDb, Sig, ViewId, Reducer),
+        GroupKeyFun = fun({Key, _DocId}) ->
+            couch_views_util:group_level_key(Key, GroupLevel)
+        end,
+
+        Acc1 = ebtree:group_reduce(Tx, Tree, StartKey, EndKey, GroupKeyFun,
+            fun fold_cb/2, Acc0, Opts),
+
+        maps:get(user_acc, Acc1)
+    end).
+
+
+get_existing_keys(ViewId, ExistingViewKeys) ->
+    case lists:keyfind(ViewId, 1, ExistingViewKeys) of
+        {ViewId, _TotalRows, _TotalSize, EKeys} ->
+            EKeys;
+        false ->
+            []
+    end.
+
+
+% The reduce values are stored as keys in the b-tree
+% So every call to the reducer from the b-tree is always
+% a rereduce call.
+get_reducer({_, ReduceFun}) ->
+    get_reducer(ReduceFun);
+
+get_reducer(Reducer) ->
+    fun
+        (Vs, true) ->
+            rereduce_val_and_size(Reducer, Vs);
+        (KVs, false) ->
+            {_, Vs} = lists:unzip(KVs),
+            rereduce_val_and_size(Reducer, Vs)
+    end.
+
+
+% This happens if a reduce is called without the index being built
+% for example from `get_kv_size`
+rereduce_val_and_size(Reducer, []) ->
+    {0, 0};
+
+rereduce_val_and_size(Reducer, Vs) ->
+    {ReduceVs, SizeVs} = lists:unzip(Vs),
+    ReduceVal = couch_views_reducer:rereduce_values(Reducer, ReduceVs),
+    SizeVal = lists:sum(SizeVs),
+    {ReduceVal, SizeVal}.
+
+
+open_tree(TxDb, Sig, ViewId, ViewReduceFun) ->
+    #{
+        db_prefix := DbPrefix,
+        tx := Tx
+    } = TxDb,
+
+    ReduceId = couch_views_util:reduce_id(ViewId, ViewReduceFun),
+    ReduceIdxPrefix = idx_prefix(DbPrefix, Sig, ReduceId),
+
+    TreeOpts = [
+        {reduce_fun, get_reducer(ViewReduceFun)}
+    ],
+    ebtree:open(Tx, ReduceIdxPrefix, btree_order_size(), TreeOpts).
+
+
+delete_keys(Tx, Tree, DocId, Keys) ->
+    lists:foreach(fun (Key) ->
+        EK = create_key(Key, DocId),
+        ebtree:delete(Tx, Tree, EK)
+    end, Keys).
+
+
+add_keys(Tx, Tree, DocId, Results) ->
+    lists:foreach(fun ({Key, Val}) ->
+        EK = create_key(Key, DocId),
+        EV = create_val(Key, Val),
+        ebtree:insert(Tx, Tree, EK, EV)
+    end, Results).
+
+
+create_key(Key, DocId) ->
+    {Key, DocId}.
+
+
+create_val(Key, Val) ->
+    KeySize = erlang:external_size(Key),
```

Comment:
if this is intended to be the billing size, it should be the `couch_ejson_size:encoded_size/1` function like currently. check with eric who changed the external size away from external_size the last time.

Reply:
I've followed how the sizing is calculated in `couch_views_indexer`. There is a note there to change to `couch_ejson_size:encoded_size/`. So we should probably change both in a separate PR.